### PR TITLE
fix(garbage collector): limit catalog update for files to delete

### PR DESCRIPTION
**Pull request by philjb**
Impose a 1000 LIMIT on flag_for_delete_by_retention so the garbage collector's load on the catalog is limited. 1000 is used as the fixed limit in another catalog DML.

* follow up to requests in #7562

Closes #

Describe your proposed changes here.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: iox_catalog/src/postgres.rs [hunk 1]
```diff
@@ -1539,17 +1539,23 @@ RETURNING *;
...
+)
+UPDATE parquet_file
+SET to_delete = $2
+WHERE id IN (SELECT id FROM parquet_file_ids)
+RETURNING id;
             "#,
         )
-        .bind(flagged_at) // $1
+        .bind(MAX_PARQUET_FILES_SELECTED_ONCE) // $1
+        .bind(flagged_at) // $2
```
Comment (philjb) [inline iox_catalog/src/postgres.rs:1558]: note the order to match the order in the DML

**Review by alamb — COMMENTED** [review]
This looks reasonable to me, but I am not really sure how to verify this does what it says. I wonder if it would be possible to add a test somehow?

**Comment by alamb** [comment]
Sorry -- let me be clear. For this PR, I think we should have some sort of catalog / unit test that shows this code works properly with more than 1000 files (namely that calling it actually limits to 1000 and that subsequent calls can successfully see the remaining files and we don't get stuck in a loop with the same 1000 files forever)

Perhaps we can extend 
https://github.com/influxdata/influxdb_iox/blob/02344af8138b45a67c286194752c9806b69a323c/iox_catalog/src/lib.rs#L238

Also, I think it is probably time to  add some end to end test coverage for the garbage collector, along the lines of what is in 

https://github.com/influxdata/influxdb_iox/tree/main/influxdb_iox/tests/end_to_end_cases

I can file a ticket for this task if that would help

**Review by domodwyer — APPROVED** [review]
**THANK YOU** it looks great! I assume the `WITH` subquery is required to LIMIT the update or something?

I agree this should have a quick unit test in the catalog crate, something simple like:

* [Create] 1,042 parquet files
* Call the method under test, and `assert_eq!(returned.len(), 1000);`
* Call the method again, and `assert_eq!(returned.len(), 42);`

That's it! You could throw in a SELECT here and there to validate the rows are being marked, but I think the decreasing return value shows it is happening too. 

Here's an existing test that creates a parquet file:

https://github.com/influxdata/influxdb_iox/blob/526a517b71c4c9d69af4a84f469fa2ca3a9a6c7a/iox_catalog/src/interface.rs#L1907

This'll let you copy/pasta the various other entities needed in the catalog before you can insert a `parquet_file` record (because of the foreign keys).

I agree some e2e tests of the garbage collector would be good in a different PR. It might be worth pairing with someone to figure out the structure of e2e tests and how they're built - they're a little different from unit tests, though there are also plenty of examples in the IOx codebase. Let me know if you want a hand :+1: 

[Create]:https://github.com/influxdata/influxdb_iox/blob/526a517b71c4c9d69af4a84f469fa2ca3a9a6c7a/iox_catalog/src/interface.rs#L551-L552

**Comment by philjb** [comment]
I will add the test!

**Comment by philjb** [comment]
> I can file a ticket for this task if that would help

pulled out to https://github.com/influxdata/influxdb_iox/issues/7646

**Comment by philjb** [comment]
@domodwyer @alamb - please take a look at the test code I added.

**Review by domodwyer — APPROVED** [review]
Perfect 👌

**Comment by philjb** [comment]
@domodwyer 

> I assume the WITH subquery is required to LIMIT the update or something?

yes - UPDATE/DELETE require a sub query/common table expression to get the LIMIT effect.

**Pull request merged**